### DISCUSSION TITLE
Add Locale and Charset to the standard library?

### DISCUSSION THREAD

**Discussion by stephentetley** [flix/flix#6357] [Ideas]
I've extracted the `Locale` and `Charset` modules from my library `interop-base` and updated them to use the new marshaling functions from `Adaptor`. I think they are good candidates for inclusion in the standard library...

Pros
We already use "stringly typed" charsets in the `File` - we could then switch to strong typing and not have to worry about typos in charset names.
`Locale` is the basis for internationalization in the JDK and is used thoughout it.
`Locale` provides a good number of examples of marshaling to and from different collection types.
We would want a single "canonical" implementation of both `Locale` and `Charset` so library writers could depend on them. If they are in the stdlib there are no dependency worries.

Cons
`Locale` is quite large and some of the functions are quite "rarefied".
`Locale` brings in `LanguageRange` which is also "rarefied" but does have a nice example of marshaling into a Map.

Code is here - comments and tests are still to write, but the implementations are pretty much complete.

https://github.com/stephentetley/charset-locale

**Comment by magnus-madsen** [comment]
Hi Stephen,

We are in this state were I am a bit hesitant to add more to the standard library. My primary concern is not yet bloat, but rather that currently the entire standard library must be loaded and compiled, regardless of whether it is used by the program. (Post type inference, we tree shake the program and then there is no cost, nor in the generated bytecode). I would like to move away from this and do a bit more compilation on demand, but this requires some changes we are not yet prepared to deal with (and is not yet our focus). So essentially I want to say neither yes or no, but rather: maybe later. I the interim I suggest to expand `Adaptor` and to keep your locale as a Flix library. Once the situation with demand-driven loading improves, I may consider including it, and if it is already a polished library then it would be easy.

As for IO/Files, I expect some rework coming there.

**Comment by stephentetley** [comment]
Thanks Magnus, I'll continue with it as a lib for now and sort of tests and docs so it can be used by others.

**Comment by mlutze** [comment]
We can also consider adding this to community build.

**Reply by magnus-madsen** [reply]
Yes, let's do it
